src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/Syntax/SyntaxTokenList.cs

```diff
@@ -0,0 +1,497 @@
+
+    public SyntaxTokenList Add(SyntaxToken token)
+        => Insert(Count, token);
+
+    public SyntaxTokenList AddRange(IEnumerable<SyntaxToken> tokens)
+        => InsertRange(Count, tokens);
+
+    public SyntaxTokenList Insert(int index, SyntaxToken token)
+    {
+        if (token == default)
+        {
+            ThrowHelper.ThrowArgumentOutOfRangeException(nameof(token));
+        }
+
+        return InsertRange(index, [token]);
+    }
+
+    public SyntaxTokenList InsertRange(int index, ReadOnlySpan<SyntaxToken> tokens)
+    {
+        var count = Count;
+
+        ArgHelper.ThrowIfNegative(index);
+        ArgHelper.ThrowIfGreaterThan(index, count);
+
+        if (tokens.Length == 0)
+        {
+            return this;
+        }
+
+        var array = new ArrayElement<GreenNode>[count + tokens.Length];
+
+        // Add current tokens up to 'index'
+        int i;
+        for (i = 0; i < index; i++)
+        {
+            array[i].Value = this[i].RequiredNode;
+        }
+
+        // Add new tokens
+        for (var j = 0; j < tokens.Length; i++, j++)
+        {
+            array[i].Value = tokens[j].RequiredNode;
+        }
+
+        // Add remaining tokens starting from 'index'
+        for (var j = index; j < count; i++, j++)
+        {
+            array[i].Value = this[j].RequiredNode;
+        }
+
+        return new(parent: null, InternalSyntax.SyntaxList.List(array), position: 0, index: 0);
+    }
+
+    public SyntaxTokenList InsertRange(int index, IEnumerable<SyntaxToken> tokens)
+    {
+        var count = Count;
+
+        ArgHelper.ThrowIfNegative(index);
+        ArgHelper.ThrowIfGreaterThan(index, count);
+        ArgHelper.ThrowIfNull(tokens);
+
+        if (tokens.TryGetCount(out var tokenCount))
+        {
+            return InsertRangeWithCount(index, tokens, tokenCount);
+        }
+
+        using var builder = new PooledArrayBuilder<SyntaxToken>(count);
+
+        // Add current tokens up to 'index'
+        for (var i = 0; i < index; i++)
+        {
+            builder.Add(this[i]);
+        }
+
+        var oldCount = builder.Count;
+
+        // Add new tokens
+        foreach (var token in tokens)
+        {
+            builder.Add(token);
+        }
+
+        // If builder.Count == oldCount, there weren't any tokens added.
+        // So, there's no need to continue.
+        if (builder.Count == oldCount)
+        {
+            return this;
+        }
+
+        // Add remaining tokens starting from 'index'
+        for (var i = index; i < count; i++)
+        {
+            builder.Add(this[i]);
+        }
+
+        return new(parent: null, builder.ToGreenListNode(), position: 0, index: 0);
+    }
+
+    private SyntaxTokenList InsertRangeWithCount(int index, IEnumerable<SyntaxToken> tokens, int tokenCount)
+    {
+        if (tokenCount == 0)
+        {
+            return this;
+        }
+
+        var count = Count;
+        var array = new ArrayElement<GreenNode>[count + tokenCount];
```

Comment:
Am I crazy, or is `tokenCount` not looked at again here? What if there are more tokens in `tokens` than `count + tokenCount`?